lib/Graph/Grad.cpp

```diff
@@ -55,7 +55,7 @@ NodeValue GraphGradMapper::getGradient(NodeValue activation) {
 
 Function *glow::differentiate(Function *F, const TrainingConfig &conf,
                               llvm::StringRef newFuncName,
-                              VariableGradientsList *varGrads) {
+                              VariableGradientsList *varGrads, Context *ctx) {
```

Comment:
Perhaps an assert that either `varGrads && ctx` or `!varGrads && !ctx` is `true` would be useful here?

Reply:
I think an assertion is a good idea.

Reply:
Perhaps we should not pass the context and instead initialize the newly created placeholders externally?

Reply:
Hm, yeah, the context isn't really doing any work here so maybe it's better to avoid passing it.  Idk.  My feelings aren't too strong on this one :).

Reply:
The newly created placeholders are allocated externally using `allocateGrads` in `gradCheckTest.cpp`. Without passing in the context, `createSave` makes a `Variable` as the output of the `SaveNode`; that's why I passed it in.

Reply:
@SplitInfinity We could create the placeholder and the SaveNode in two builder calls. Alternatively we could create a new API: createSaveNodePH.